### PR TITLE
Persistent session

### PR DESCRIPTION
**Pull request by g-krisztian**
### Description
When app's configuration  
```clojure
(= :database (get-in config [:framework.app/session-backend :storage])) 
```
then it uses the app database (by default) to store session data

Options in `:framework.app/session-backend` keys:
 - `:session-table-name` to store session, `:sessions` by default
 - `{:port     5433
     :dbname   "frankie"
     :host     "localhost"
     :dbtype   "postgresql"
     :user     "flexiana"
     :password "dev"}` to specify different storage
 - if initialized after db connection, using the same datasource
 - if no `:db :datasource` provided, merges app's database configuration with session-backend's and creates another datasource.

### Tester info
try it with Frankie from https://github.com/Flexiana/Frankie/tree/persistent-session branch.
 - login
 - restart the backend-app
 - refresh page shouldn't forward to login screen.

### CODE REVIEW

**Review by soulflyer — APPROVED** [review]
Nice work!

File: doc/How-To.md [hunk 1]
```diff
@@ -209,6 +210,82 @@ and use it:
...
+    (-> (config/config app-cfg)
+        routes/reset
+        db-core/connect
+        db-core/migrate!
+        session/init-backend
+        ws/start))
+  ```
+    - Creating new datasource
+
+  If no datasource provided on initialization, the `init-backend` function merging the database config with session
```
Comment (soulflyer) [inline doc/How-To.md:283]: "If no datasource is provided on initialization, the `init-backend` function merges the database config with the session backend configuration, and creates a new datasource from the result."

**Comment by Yacobh** [comment]
testing controller example I put 
```
:framework.app/session-backend
 {:storage            :database
  :session-table-name :sessions
  :port               5432
  :dbname             "controllers"
  :host               "localhost"
  :dbtype             "postgresql"
  :user               "postgres"
  :password           "postgres"}
```

this code in the config.edn file and

```
(defn ->system
  [app-cfg]
  (-> (config/config)
      (merge app-cfg)
      routes/reset
      session/init-backend
      ws/start
      closeable-map))
```

and this other in the core.clj 
then I run `lein test` 

and get this error:

```
lein test core-test

lein test :only core-test/core-test

ERROR in (core-test) (core_deftype.clj:583)
Uncaught exception in test fixture
expected: nil
  actual: java.lang.IllegalArgumentException: No implementation of method: :get-datasource of protocol: #'next.jdbc.protocols/Sourceable found for class: nil
 at clojure.core$_cache_protocol_fn.invokeStatic (core_deftype.clj:583)
    clojure.core$_cache_protocol_fn.invoke (core_deftype.clj:575)
    next.jdbc.protocols$eval14946$fn__14947$G__14937__14952.invoke (protocols.clj:14)
    next.jdbc$get_datasource.invokeStatic (jdbc.clj:132)
    next.jdbc$get_datasource.invoke (jdbc.clj:69)
    framework.db.core$get_datasource.invokeStatic (core.clj:40)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:42)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.db.core$get_datasource.invokeStatic (core.clj:33)
    framework.db.core$get_datasource.invoke (core.clj:31)
    framework.session.core$init_in_db.invokeStatic (core.clj:72)
    framework.session.core$init_in_db.invoke (core.clj:68)
    framework.session.core$init_backend.invokeStatic (core.clj:223)
    framework.session.core$init_backend.invoke (core.clj:217)
    core$__GT_system.invokeStatic (core.clj:15)
    core$__GT_system.invoke (core.clj:13)
    core_test$std_system_fixture.invokeStatic (core_test.clj:10)
    core_test$std_system_fixture.invoke (core_test.clj:8)
    clojure.test$compose_fixtures$fn__9731$fn__9732.invoke (test.clj:694)
    clojure.test$default_fixture.invokeStatic (test.clj:687)
    clojure.test$default_fixture.invoke (test.clj:683)
    clojure.test$compose_fixtures$fn__9731.invoke (test.clj:694)
    clojure.test$test_vars.invokeStatic (test.clj:731)
    clojure.test$test_all_vars.invokeStatic (test.clj:737)
    clojure.test$test_ns.invokeStatic (test.clj:758)
    clojure.test$test_ns.invoke (test.clj:743)
    user$eval224$fn__287.invoke (form-init13885357636145950991.clj:1)
    clojure.lang.AFn.applyToHelper (AFn.java:156)
    clojure.lang.AFn.applyTo (AFn.java:144)
    clojure.core$apply.invokeStatic (core.clj:667)
    clojure.core$apply.invoke (core.clj:660)
    leiningen.core.injected$compose_hooks$fn__154.doInvoke (form-init13885357636145950991.clj:1)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:665)
    clojure.core$apply.invoke (core.clj:660)
    leiningen.core.injected$run_hooks.invokeStatic (form-init13885357636145950991.clj:1)
    leiningen.core.injected$run_hooks.invoke (form-init13885357636145950991.clj:1)
    leiningen.core.injected$prepare_for_hooks$fn__159$fn__160.doInvoke (form-init13885357636145950991.clj:1)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:31)
    clojure.lang.RestFn.invoke (RestFn.java:408)
    clojure.core$map$fn__5866.invoke (core.clj:2755)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:51)
    clojure.lang.Cons.next (Cons.java:39)
    clojure.lang.RT.boundedLength (RT.java:1792)
    clojure.lang.RestFn.applyTo (RestFn.java:130)
    clojure.core$apply.invokeStatic (core.clj:667)
    clojure.test$run_tests.invokeStatic (test.clj:768)
    clojure.test$run_tests.doInvoke (test.clj:768)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:665)
    clojure.core$apply.invoke (core.clj:660)
    user$eval224$fn__297$fn__330.invoke (form-init13885357636145950991.clj:1)
    user$eval224$fn__297$fn__298.invoke (form-init13885357636145950991.clj:1)
    user$eval224$fn__297.invoke (form-init13885357636145950991.clj:1)
    user$eval224.invokeStatic (form-init13885357636145950991.clj:1)
    user$eval224.invoke (form-init13885357636145950991.clj:1)
    clojure.lang.Compiler.eval (Compiler.java:7177)
    clojure.lang.Compiler.eval (Compiler.java:7167)
    clojure.lang.Compiler.load (Compiler.java:7636)
    clojure.lang.Compiler.loadFile (Compiler.java:7574)
    clojure.main$load_script.invokeStatic (main.clj:475)
    clojure.main$init_opt.invokeStatic (main.clj:477)
    clojure.main$init_opt.invoke (main.clj:477)
    clojure.main$initialize.invokeStatic (main.clj:508)
    clojure.main$null_opt.invokeStatic (main.clj:542)
    clojure.main$null_opt.invoke (main.clj:539)
    clojure.main$main.invokeStatic (main.clj:664)
    clojure.main$main.doInvoke (main.clj:616)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.lang.Var.applyTo (Var.java:705)
    clojure.main.main (main.java:40)

Ran 0 tests containing 1 assertions.
0 failures, 1 errors.
Tests failed.
```

**Comment by g-krisztian** [comment]
> testing controller example I put
> 
> ```
> :framework.app/session-backend
>  {:storage            :database
>   :session-table-name :sessions
>   :port               5432
>   :dbname             "controllers"
>   :host               "localhost"
>   :dbtype             "postgresql"
>   :user               "postgres"
>   :password           "postgres"}
> ```
> 
> this code in the config.edn file and
> 
> ```
> (defn ->system
>   [app-cfg]
>   (-> (config/config)
>       (merge app-cfg)
>       routes/reset
>       session/init-backend
>       ws/start
>       closeable-map))
> ```
> 
> and this other in the core.clj then I run `lein test`
> 
> and get this error:
> 
> ```
> lein test core-test
> 
> lein test :only core-test/core-test
> 
> ERROR in (core-test) (core_deftype.clj:583)
> Uncaught exception in test fixture
> expected: nil
>   actual: java.lang.IllegalArgumentException: No implementation of method: :get-datasource of protocol: #'next.jdbc.protocols/Sourceable found for class: nil
>  at clojure.core$_cache_protocol_fn.invokeStatic (core_deftype.clj:583)
>     clojure.core$_cache_protocol_fn.invoke (core_deftype.clj:575)
>     next.jdbc.protocols$eval14946$fn__14947$G__14937__14952.invoke (protocols.clj:14)
>     next.jdbc$get_datasource.invokeStatic (jdbc.clj:132)
>     next.jdbc$get_datasource.invoke (jdbc.clj:69)
>     framework.db.core$get_datasource.invokeStatic (core.clj:40)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:42)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.db.core$get_datasource.invokeStatic (core.clj:33)
>     framework.db.core$get_datasource.invoke (core.clj:31)
>     framework.session.core$init_in_db.invokeStatic (core.clj:72)
>     framework.session.core$init_in_db.invoke (core.clj:68)
>     framework.session.core$init_backend.invokeStatic (core.clj:223)
>     framework.session.core$init_backend.invoke (core.clj:217)
>     core$__GT_system.invokeStatic (core.clj:15)
>     core$__GT_system.invoke (core.clj:13)
>     core_test$std_system_fixture.invokeStatic (core_test.clj:10)
>     core_test$std_system_fixture.invoke (core_test.clj:8)
>     clojure.test$compose_fixtures$fn__9731$fn__9732.invoke (test.clj:694)
>     clojure.test$default_fixture.invokeStatic (test.clj:687)
>     clojure.test$default_fixture.invoke (test.clj:683)
>     clojure.test$compose_fixtures$fn__9731.invoke (test.clj:694)
>     clojure.test$test_vars.invokeStatic (test.clj:731)
>     clojure.test$test_all_vars.invokeStatic (test.clj:737)
>     clojure.test$test_ns.invokeStatic (test.clj:758)
>     clojure.test$test_ns.invoke (test.clj:743)
>     user$eval224$fn__287.invoke (form-init13885357636145950991.clj:1)
>     clojure.lang.AFn.applyToHelper (AFn.java:156)
>     clojure.lang.AFn.applyTo (AFn.java:144)
>     clojure.core$apply.invokeStatic (core.clj:667)
>     clojure.core$apply.invoke (core.clj:660)
>     leiningen.core.injected$compose_hooks$fn__154.doInvoke (form-init13885357636145950991.clj:1)
>     clojure.lang.RestFn.applyTo (RestFn.java:137)
>     clojure.core$apply.invokeStatic (core.clj:665)
>     clojure.core$apply.invoke (core.clj:660)
>     leiningen.core.injected$run_hooks.invokeStatic (form-init13885357636145950991.clj:1)
>     leiningen.core.injected$run_hooks.invoke (form-init13885357636145950991.clj:1)
>     leiningen.core.injected$prepare_for_hooks$fn__159$fn__160.doInvoke (form-init13885357636145950991.clj:1)
>     clojure.lang.RestFn.applyTo (RestFn.java:137)
>     clojure.lang.AFunction$1.doInvoke (AFunction.java:31)
>     clojure.lang.RestFn.invoke (RestFn.java:408)
>     clojure.core$map$fn__5866.invoke (core.clj:2755)
>     clojure.lang.LazySeq.sval (LazySeq.java:42)
>     clojure.lang.LazySeq.seq (LazySeq.java:51)
>     clojure.lang.Cons.next (Cons.java:39)
>     clojure.lang.RT.boundedLength (RT.java:1792)
>     clojure.lang.RestFn.applyTo (RestFn.java:130)
>     clojure.core$apply.invokeStatic (core.clj:667)
>     clojure.test$run_tests.invokeStatic (test.clj:768)
>     clojure.test$run_tests.doInvoke (test.clj:768)
>     clojure.lang.RestFn.applyTo (RestFn.java:137)
>     clojure.core$apply.invokeStatic (core.clj:665)
>     clojure.core$apply.invoke (core.clj:660)
>     user$eval224$fn__297$fn__330.invoke (form-init13885357636145950991.clj:1)
>     user$eval224$fn__297$fn__298.invoke (form-init13885357636145950991.clj:1)
>     user$eval224$fn__297.invoke (form-init13885357636145950991.clj:1)
>     user$eval224.invokeStatic (form-init13885357636145950991.clj:1)
>     user$eval224.invoke (form-init13885357636145950991.clj:1)
>     clojure.lang.Compiler.eval (Compiler.java:7177)
>     clojure.lang.Compiler.eval (Compiler.java:7167)
>     clojure.lang.Compiler.load (Compiler.java:7636)
>     clojure.lang.Compiler.loadFile (Compiler.java:7574)
>     clojure.main$load_script.invokeStatic (main.clj:475)
>     clojure.main$init_opt.invokeStatic (main.clj:477)
>     clojure.main$init_opt.invoke (main.clj:477)
>     clojure.main$initialize.invokeStatic (main.clj:508)
>     clojure.main$null_opt.invokeStatic (main.clj:542)
>     clojure.main$null_opt.invoke (main.clj:539)
>     clojure.main$main.invokeStatic (main.clj:664)
>     clojure.main$main.doInvoke (main.clj:616)
>     clojure.lang.RestFn.applyTo (RestFn.java:137)
>     clojure.lang.Var.applyTo (Var.java:705)
>     clojure.main.main (main.java:40)
> 
> Ran 0 tests containing 1 assertions.
> 0 failures, 1 errors.
> Tests failed.
> ```

Addressed in the new commit. thank you @Yacobh and @puchka

**Review by simionion — APPROVED** [review]
All works as expected.

**Review by Yacobh — APPROVED** [review]
It works well

**Pull request merged**